File: blacklist.pl
#!/usr/bin/perl
use CGI qw/:standard -no_xhtml/;
use URI::Escape;
use strict;
use oyster::conf;
use oyster::common;

my %config = oyster::conf->get_config('oyster.conf');

oyster::common->navigation_header();

my $savedir = $config{'savedir'};
my $basedir = $config{'basedir'};
my $cssclass = 'file2';
my @results = ();

my $affects = '';
if (param('affects') && (param('action') eq 'test')) {
    $affects = param('affects');
}

$affects =~ s/&/&amp;/g;

print "<form action='blacklist.pl'>";
print "<table><tr>";
print "<td><input type='text' name='affects' value='$affects'></td>";
print "<td><input type='radio' name='action' value='test' checked> Test only<br>";
print "<input type='radio' name='action' value='add'> Add to Blacklist</td>";
print "<td style='padding-left: 2em;'><input type='submit' value='Go'></td></tr></table>\n</form>\n";

print p("<a href='blacklist.pl'>Show current blacklist</a>");

if (param('action') && param('affects')) {
    if (param('action') eq 'test') {
	print_affects(param('affects'));
    } elsif (param('action') eq 'add') {
	add_to_blacklist(param('affects'));
	print_blacklist();
    } elsif (param('action') eq 'delete') {
	delete_from_blacklist(param('affects'));
	print_blacklist();
    }
} else {
    print_blacklist();
}

print end_html;

exit 0;

sub print_blacklist {

    # Opens current blacklist and prints each line

    my @blacklistlines = ();
    open (FILE, "${savedir}blacklist");
    while (my $line = <FILE>) {
	chomp($line);
	push (@blacklistlines, $line);
    }
    close (FILE);

    open (LIST, "${savedir}lists/default");

    my $mediadir = $config{'mediadir'};
    $mediadir =~ s/\/$//;

    my %lineaffects = ();

    # Count affected files for each rule

    while (my $line = <LIST>) {
	chomp($line);
	$line =~ s/^\Q$mediadir\E//;
	foreach my $blacklistline (@blacklistlines) {
	    if ($line =~ /$blacklistline/i) {
		if ($lineaffects{$blacklistline}) {
		    $lineaffects{$blacklistline}++;
		} else {
		    $lineaffects{$blacklistline} = 1;
		}
	    }
	}
    }
    close (LIST);

    my $totalaffected = 0;

    print "<table width='100%'>";
    foreach my $line (@blacklistlines) {
	$totalaffected += $lineaffects{$line};
	my $escapedline = uri_escape("$line", "^A-Za-z");
	print "<tr><td width='60%'>$line</td>";
	print "<td width='25%' align='left'><a href='blacklist.pl?action=test&amp;affects=$escapedline'>Affects</a> ($lineaffects{$line})</td>";
	print "<td width='15%' align='center'><a href='blacklist.pl?action=delete&amp;affects=$escapedline'>Delete</a></td></tr>";
    }

    print "</table>\n";

    print p(strong("Total files affected:"), $totalaffected);
}

sub print_affects {

    # Shows all files, which are affected by a blacklist-rule

    my $affects = $_[0];
    open (LIST, "${savedir}lists/default");

    # Add all matching lines to @results

    my $mediadir = $config{'mediadir'};
    $mediadir =~ s/\/$//;

    while (my $line = <LIST>) {
	chomp($line);
	$line =~ s/^\Q$mediadir\E//;
	if ($line =~ /$affects/i) {
	    push (@results, $line);
	}
    }
    close (LIST);

    # Sort @results alphabetically

    @results = sort @results;

    # Determine maximum depth of directories for further sorting

    my $maxdepth = 0;
    foreach my $result (@results) {
	my $line = $result;
	my $counter = 0;
	while ($counter < $maxdepth) {
	    $line =~ s/^[^\/]*\///;
	    $counter++;
	}
	if ($line =~ /\//) {
	    $maxdepth++;
	}
    }
    $maxdepth--;

    # Sort @results by a given depth

    while ($maxdepth >= 0) {
	@results = sort_results($maxdepth);
	$maxdepth--;
    }

    if (@results > 0) {
	listdir('/', 0);
    } else {
	print p('No songs match these rule.');
    }
}

sub add_to_blacklist {

    # Appends a rule to the blacklist

    my $affects = $_[0];
    open (BLACKLIST, ">>${savedir}blacklist");
    print BLACKLIST "$affects\n";
    close (BLACKLIST);
}

sub delete_from_blacklist {

    # removes a rule from the blacklist

    my $affects = $_[0];
    system ("cp ${savedir}blacklist ${basedir}blacklist.tmp");
    open (BLACKLIST, "${basedir}blacklist.tmp");
    open (NEWBLACKLIST, ">${savedir}blacklist");
    while (my $line = <BLACKLIST>) {
	if (!($line =~ /^\Q$affects\E$/)) {
	    print NEWBLACKLIST $line;
	}
    }
    close (BLACKLIST);
    close (NEWBLACKLIST);
    system ("rm -f ${basedir}blacklist.tmp");
}

sub listdir {

    # lists the directory $basepath and recursive all subdirs

    my $basepath = $_[0];
    my $counter = $_[1];

    while (($counter < @results) && ($results[$counter] =~ /^\Q$basepath\E/)) {
	my $newpath = $results[$counter];
	$newpath =~ s/^\Q$basepath\E//;
	if ($newpath =~ /\//) {
	    $newpath =~ /^([^\/]*\/)/;
	    $newpath = $1;

	    my $cutnewpath = $newpath;
	    $cutnewpath =~ s/\/$//;
	    my $escapedcutnewpath = $cutnewpath;
	    $escapedcutnewpath =~ s/&/&amp;/g;

	    if (!($basepath eq '/')) {
		my $escapeddir = uri_escape("$basepath$cutnewpath", "^A-Za-z");
		print "<div style='padding-left: 1em;'><strong><a href='browse.pl?dir=$escapeddir'>$escapedcutnewpath</a></strong>\n";
		$newpath = "$basepath$newpath";
	    }  else {
		my $escapeddir = uri_escape("$cutnewpath", "^A-Za-z");
		print "<strong><a href='browse.pl?dir=$escapeddir'>$escapedcutnewpath</a></strong>";
		$newpath = "/$newpath";
	    }
	    $counter = listdir($newpath,$counter);
	     if (!($basepath eq '/')) {
		print "</div>\n";
	    }
	} else {
	    print "<div style='padding-left: 1em;'>";
	    while ($results[$counter] =~ /^\Q$basepath\E/) {
		my $filename = $results[$counter];
		$filename =~ s/^.*\///;
		$filename =~ /(.*)\.(...)$/;
		my $nameonly = $1;
		$nameonly =~ s/&/&amp;/g;
		my $escapedfile = uri_escape("$basepath$filename", "^A-Za-z");
		if ($cssclass eq 'file') {
		    $cssclass = 'file2';
		} else {
		    $cssclass = 'file';
		}
		print "<a href='fileinfo.pl?file=$escapedfile' class='$cssclass'>$nameonly</a><br>\n";
		$counter++;
	    }
	    print "</div>";
	}
    }

    return ($counter);

}

sub sort_results {

    # sorts @results by a given directory depth
    # directories first, then regular files

    my $depth = $_[0];
    my (@dirs, @files) = ();

    foreach my $result (@results) {
	my $line = $result;
	my $counter = $depth;
	while ($counter > 0) {
	    $line =~ s/^[^\/]*\///;
	    $counter--;
	}
	if ($line =~ /\//) {
	    push (@dirs, $result);
	} else {
	    push (@files, $result);
	}
    }

    return (@dirs, @files);

}

File: blacklist.py
#!/usr/bin/python
# -*- coding: UTF-8 -*-

# oyster - a python-based jukebox and web-frontend
#
# Copyright (C) 2004 Benjamin Hanzelmann <ben@nabcos.de>,
#  Stephan Windmüller <windy@white-hawk.de>,
#  Stefan Naujokat <git@ethric.de>
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

"""
Allows the user to specify regex-lines for files
which should not be played at random
"""

__revision__ = 1

import cgi
import config
import cgitb
import os.path
import urllib
import common
import re
cgitb.enable()

def print_blacklist ():

    "Opens current blacklist and prints each line"

    lineaffects = {}
    blacklistmatchers = {}

    if os.path.exists(myconfig['savedir'] + "blacklists/" + playlist):
        blacklist = open(myconfig['savedir'] + "blacklists/" + playlist)
        for line in blacklist.readlines():
            blacklistmatchers[line[:-1]] = re.compile(line.rstrip())
            line = line.replace(mediadir, '', 1)[:-1]
            lineaffects[line] = 0
        blacklist.close()

    listfile = open(myconfig['savedir'] + "lists/" + playlist)

    totalaffected = 0

    # Count affected files for each rule

    for line in listfile.readlines():
        isblacklisted = 0
        line = line.replace(mediadir, '', 1)[:-1]
        for key in blacklistmatchers.keys():
            if blacklistmatchers[key].search(line):
                isblacklisted = 1
                lineaffects[key] += 1
        if isblacklisted:
            totalaffected += 1

    listfile.close()

    blacklistlines = blacklistmatchers.keys()
    blacklistlines.sort()
    print "<table width='100%'>"
    for line in blacklistlines:
        escapedline = urllib.quote(line)
        print "<tr><td width='60%'>" + line + "</td>"
        print "<td width='25%' align='left'><a href='blacklist.py?" + \
            "action=test&amp;affects=" + escapedline + "'>Affects</a> (" + \
            str(lineaffects[line]) + ")</td>"
        print "<td width='15%' align='center'><a href='blacklist.py?" + \
            "action=delete&amp;affects=" + escapedline + \
            "'>Delete</a></td></tr>"

    print "</table>"

    print "<p><strong>Total files affected:</strong> " + \
        str(totalaffected) + "</p>"

def print_affects (affects):

    "Shows all files, which are affected by a blacklist-rule"

    affectresults = []

    # Add all matching lines to results

    if os.path.exists(myconfig['savedir'] + "lists/" + playlist):
        listfile = open (myconfig['savedir'] + "lists/" + playlist, 'r')
        for line in listfile.readlines():
            line = line.replace(mediadir, '', 1)[:-1]
            if re.compile(affects).search(line) != None:
                affectresults.append(line)
        listfile.close()

    # Sort results alphabetically

    if affectresults != []:
        affectresults.sort()
        common.results = affectresults
        common.listdir('/', 0,'file2')
    else:
        print "<p>No songs match these rule.</p>"

def add_to_blacklist (affects):

    "Appends a rule to the blacklist"

    blacklist = open(savedir + "blacklists/" + playlist, 'a')
    blacklist.write(affects + "\n")
    blacklist.close()

def delete_from_blacklist (affects):

    "removes a rule from the blacklist"

    os.system ("cp \"" + savedir + "blacklists/" + playlist + "\" " + \
        savedir + "blacklist.tmp")
    blacklist = open(savedir + "blacklist.tmp")
    newblacklist = open(savedir + "blacklists/" + playlist, 'w')
    for line in blacklist.readlines():
        if line[:-1] != affects:
            newblacklist.write(line)
    blacklist.close()
    newblacklist.close()
    os.unlink (savedir + "blacklist.tmp")

myconfig = config.get_config()
basedir = myconfig['basedir']
savedir = myconfig['savedir']
mediadir = myconfig['mediadir'][:-1]
form = cgi.FieldStorage()
playlist = config.get_playlist()

import mCommon
mCommon.navigation_header("Blacklists")

results = []

if form.has_key('affects') and form.has_key('action') \
    and form['action'].value == 'test':
    escaffects = cgi.escape(form['affects'].value)
else:
    escaffects = ''

# Create form

print "<form method='post' action='blacklist.py' " + \
    "enctype='application/x-www-form-urlencoded'>"
print "<fieldset class='searchform'>"
print "<legend class='searchform'>Add entries to blacklist</legend>"
print "<input id='searchfield' type='text' size='40' name='affects' value='" + escaffects + "'/>"
print "<input id='searchsubmit' type='submit' name='.submit' value='Go'/>"
print "<table id='searchoptions'><tr>"
print "<td><input type='radio' name='action' value='test' checked='checked'/> " + \
    "Test only</td></tr>"
print "<tr><td><input type='radio' name='action' value='add'/> " + \
    "Add to blacklist</td></tr>"
print "</table>"
print "</fieldset>"
print "</form>"

print "<p><a href='blacklist.py'>Show current blacklist</a></p>"

if form.has_key('action') and form.has_key('affects'):
    if form['action'].value == 'test':
        print_affects(form['affects'].value)
    elif form['action'].value == 'add':
        add_to_blacklist(form['affects'].value)
        print_blacklist()
    elif form['action'].value == 'delete':
        delete_from_blacklist(form['affects'].value)
        print_blacklist()
else:
    print_blacklist()

print "</body></html>"
